src/client/Microsoft.Identity.Client/Cache/CacheKeyFactory.cs

```diff
@@ -78,17 +83,25 @@ private static bool GetOboOrAppKey(AuthenticationRequestParameters requestParame
                 requestParameters.ApiId == ApiEvent.ApiIds.AcquireTokenForUserAssignedManagedIdentity)
             {
                 string tenantId = requestParameters.Authority.TenantId ?? "";
-                key = GetClientCredentialKey(requestParameters.AppConfig.ClientId, tenantId, requestParameters.AuthenticationScheme?.KeyId);
-
+                key = GetTokenCacheItemKey(requestParameters.AppConfig.ClientId, tenantId, requestParameters.AuthenticationScheme?.KeyId, requestParameters.CacheKeyComponents);
                 return true;
             }
 
             key = null;
             return false;
         }
 
-        public static string GetClientCredentialKey(string clientId, string tenantId, string popKid)
+        public static string GetTokenCacheItemKey(
+            string clientId,
+            string tenantId,
+            string popKid,
+            SortedDictionary<string, string> cacheKeyComponents = null)
         {
+            if (cacheKeyComponents != null && cacheKeyComponents.Any())
+            {
+                return $"{popKid}{clientId}_{tenantId}_AppTokenCache_{CoreHelpers.ComputeAccessTokenExtCacheKey(cacheKeyComponents)}";
```

Comment:
Please add this after the {tenantId} part, to keep the "_AppTokenCache" suffix as is.